README.md
# QNapi [![Build Status](https://travis-ci.org/QNapi/qnapi.svg?branch=master)](https://travis-ci.org/QNapi/qnapi) [![Build status](https://ci.appveyor.com/api/projects/status/mdebep2uhrc3oagq/branch/master?svg=true)](https://ci.appveyor.com/project/krzemin/qnapi/branch/master) [![GitHub release](https://img.shields.io/github/release/QNapi/qnapi.svg)](https://github.com/QNapi/qnapi/releases) [![Github All Releases](https://img.shields.io/github/downloads/QNapi/qnapi/total.svg)](https://qnapi.github.io/#download)

QNapi is free software for automatic fetching subtitles for given movie file.
It uses online databases such as NapiProjekt, OpenSubtitles and Napisy24.
It is based on Qt5 library, so it can be launched on any supported operating
system, including Windows, OSX and Linux.

## Binary packages

Latest binary packages are available at http://qnapi.github.io/#download

## Building from source

#### Prerequisites

* C++ compiler with c++11 support installed (`clang++`, `g++` or *MinGW* for Windows), present in `PATH`
* Qt 5.2+ (most recent 5.x recommended) installed with `qmake` present in `PATH` (you can find one at http://www.qt.io/download-open-source/)

#### Binary prerequisites

QNapi requires these binary dependencies:

* p7zip (7z, 7za) - to unpack subtitles, which are commonly compressed with 7zip
* libmediainfo - to retrieve movie info such as dimensions, duration and frame rate

Linux/UNIX users can find these dependencies in separate packages.

Statically compiled **p7zip** binaries are provided in this repository for Windows/OSX
users at `win32/content/7za.exe` and `macx/content/7za`, respectively. Similarly,
compiled **libmediainfo** libraries are provided for Windows/OSX in
`deps/libmediainfo/`.

> **WARNING!** Precompiled binaries are stripped from the source archive!

#### Cloning the source code

First, you have to clone project source code using git client:

`$ git clone --recursive https://github.com/QNapi/qnapi.git`

#### Compiling

To compile the application, you have to execute two following commands in `qnapi` root directory:

`$ qmake`

This will produce `Makefile`.

`$ make` (or `mingw32-make` on Windows)

This will compile the sources and build executable binary (or app bundle on OSX).

> **Important!** Windows users have to execute one more command:
>
> `$ make install` (or `mingw32-make install`)
>
> This one will copy all binaries, libraries and other dependencies to `win32/out` directory.

#### Running

By default, output binaries are placed by `make` in different locations, depending on your operating system:

* Linux - `qnapi` in root project directory
* OSX - `macx/QNapi.app` bundle
* Windows - `win32/out/qnapi.exe` executable

After you locate your binaries, you can run the application.

> ##### 7zip note #####
> For proper subtitle extraction after download, *7zip* executable is required to be passed in application's settings.
> Linux users have to install 7zip binary package from distribution repositories or compile on its own.
> For Windows and OSX there are pre-built binaries included in this repository, in `win32` and `macx` directories appropriately and should be automatically detected by the application.

## Making redistributable package

### OSX

#### Prerequisites

You need `appdmg` script installed. You can found it at https://github.com/LinusU/node-appdmg

#### Building .dmg image

To build .dmg image for OSX with nice, drag&drop installer, you have to execute:

`$ make appdmg`

`QNapi-x.y.z.dmg` will appear in `macx` directory when command is completed.

### Windows

#### Prerequisites

You need to have **NSIS** 2.x installed. You can found it at http://nsis.sourceforge.net

Also, you will need to manually install NSIS plugin **nsProcess**. It can be found at
http://nsis.sourceforge.net/NsProcess_plugin

#### Building Windows installer

Installer script is placed at `win32/QNapi-setup.nsi`. You can build binary exe package using NSIS user interface (by `right mouse button -> compile NSIS script`) or from command line:

`$ C:\Path\To\makensis.exe QNapi-setup.nsi`

After a while, `QNapi-x.y.z-setup.exe` file will appear in `win32` directory.


doc/ChangeLog
QNapi ChangeLog:
================

2017-05-?? v0.2.3:
 * mozliwosc wyboru jezyka interfejsu uzytkownika (angielski/polski/włoski) - (thx Fincer, luigino)
 * przebudowany podsystem konfiguracji aplikacji
 * podzial aplikacji na biblioteke (libqnapi) i interfejs uzytkownika (gui)
 * wybor docelowego formatu/rozszerzenia napisow przez argument wiersza polecen - #65
 * naprawiony problem z utrata informacji o kolorze napisow podczas konwersji - #100
 * binarna wersja na Linuksa x86_64 dostepna jako AppImage

2016-10-22 v0.2.2:
 * odswiezona ikona aplikacji
 * wyswietlanie jezyka pobranych napisow w oknie podsumowania - #53 (thx andrewpros)
 * poprawki w instalatorze dla Windows - #55
 * usprawnienia w oknie wyboru napisow - #57 (thx andrewpros)
 * poprawka bledu przy zamianie znakow diakrytycznych na ASCII - #64
 * zapobieganie segfaultowi - #83 (thx kempniu)
 * poprawki literowek w kilku miejscach - #84 (thx kempniu)
 * konfiguracja silnika Napisy24 - #88
 * usuniete niewspierane funkcje silnika Napiprojekt (upload napisow,
   zglaszanie poprawek) 

2016-01-06 v0.2.1:
 * problem z nazwami sciezek zawierajacymi znaki specjalne poprawiony rowniez
   gdy otwarta jest inna instancja aplikacji (Windows) - #28
 * poprawki w detekcji kodowania UTF-8 dla jezykow innych niz polski - #37
 * rozszerzenie pliku nie jest zmieniane, gdy przetwarzanie napisow jest nieaktywne - #38
 * pokazywanie informacje o wersji Qt
 * pobieranie napisow w zapasowym jezyku w trybie cichym - #42
 * uzycie libmediainfo zamiast ffprobe do pobierania informacji o pliku filmowym
 * opcja wyszukiwania wszystkich napisow od razu, takze w jezyku zapasowym - #45
 * poprawne przerywanie wyszukiwania tylko, jesli znaleziono napisy - #46
 * poprawki bledow w silniku Napisy24 - #47


2015-11-22 v0.2.0:
 * możliwość wyboru cichego trybu wsadowego - #13
 * poprawki bledow w silniku Napisy24 - #23
 * umieszczenie ikon w katalogach opisujacych ich rozmiar - #24 (thx Fisiu)
 * wsparcie dla ekranow retina (OSX) - #27
 * wsparcie dla nazw sciezek zawierajacych znaki specjalne (Windows) - #28
 * poprawiony blad przy konwersji SRT polegajacy na dodawaniu pustych linii (Windows) - #29
 * automatyczne zamykanie aplikacji po pobraniu napisow (OSX) - #31
 * informacje o wersji w pliku .exe (Windows) - #32
 * poprawki bledow w wykrywaniu kodowania pliku (m.in. UTF-8 z BOM) - #34
 * zachowanie kodowania znakow podczas konwersji pomiedzy formatami
 * poprawiony blad nieuwzgledniajacy jezyka zapasowego przy pobieraniu napisow - #36

2015-08-26 v0.1.9:
 * obsluga konwersji pomiedzy formatami napisow: SRT, mDVD, TMP, MPL2
   (wymagany jest program ffprobe) - #10
 * mozliwosc wyboru rozszerzenia dopasowanych napisow
 * skanowanie katalogow uwzglednia ignorowanie innych rozszerzen napisow
   niz .txt (.srt i .sub) - #19
 * uwzglednianie INSTALL_PREFIX przy instalacji ikon i plikow .desktop - #20, #21

2015-08-08 v0.1.8:
 * mozliwosc wskazania zapasowego jezyka pobierania napisow
 * poprawki bledow w silniku Napisy24 dla Windows
 * poprawiona integracja z menu kontekstowym na Windows
 * wersja przenosna nie wymagajaca instalacji (archiwum zip dla Windows)
 * tryb cichy (-q) dostepny takze pod Windows
 * poprawiony blad z rozmieszczeniem okien przy wielu ekranach

2015-07-25 v0.1.7:
 * silnik dla Napisy24
 * migracja do Qt5
 * zamiana znakow diakrytycznych na odpowiedniki ASCII
 * poprawki interfejsu w trybach GUI i CLI
 * poprawki wielu innych znalezionych bledow

2009-03-04 v0.1.6 (rc2):
 * poprawki znalezionych bledow
 * kompilacja z Qt 4.5 dla Mac OS X, dzieki czemu brak ikony w doku
 * kosmetyczne poprawki interfejsu

2009-02-16 v0.1.6 (rc1):
 * wsparcie dla roznych silnikow pobierania napisow, w tym eksperymentalna
   obsluga serwisu opensubtitles.org
 * usprawnione przetwarzanie napisow
 * skanowanie katalogu zadanego jako argument wywolania aplikacji
 * liczne ulepszenia w skanowaniu katalogow
 * przywracanie domyslnych ustawien
 * znacznie rozbudowany tekstowy interfejs programu
 * natywne okna dialogowe pod Mac OS X i Windows
 * usuniete znalezione bledy, dodane nowe

2008-05-22 v0.1.5:
 * uregulowane kwestie licencyjne; od tej chwili program wydawany jest na licencji
   GNU General Public License w wersji 2. lub pozniejszej (thx Patryk Cisek)
 * usuniety blad powodujacy niemoznosc uruchomienia programu na terminalu tekstowym (TTY)
 * poprawione wysylanie napisow do serwera
 * sprawdzanie uprawnien zapisu do katalogu przy pobieraniu napisow
 * zunifikowane opcje pozwalajace na wykonywanie przetwarzania pobranych napisow
   (m.in usuwanie linii zawierajacych podane slowa, zmiana uprawnien do pliku)
 * bardziej szczegolowe podsumowanie pobierania napisow
 * mnostwo zmian w kodzie programu umozliwiajacych dalszy, swobodny rozwoj
 * sporo usprawnien w interfejsie uzytkownika
 * automatyczna (na razie eksperymentalna) integracja z GNOME (Nautilus)

2008-04-08 v0.1.4:
 * rejestracja nowych uzytkownikow za pomoca programu
 * port na systemy Windows
 * sprawdzanie poprawnosci ustawien p7zip oraz katalogu tymczasowego
 * bardziej rozbudowane podsumowanie pobierania plikow w niektorych przypadkach
 * w danej chwili moze byc uruchomiona tylko jedna instancja programu
 * liczne usprawnienia w interfejsie uzytkownika
 * pobieranie wielu plikow naraz po przeciagnieciu ich na ikone programu (Mac OS X)
 * obsluga drag-n-drop na okno pobierania (na razie nie dziala pod Mac OS X)
 * automatyczna (na razie eksperymentalna) integracja z KDE (Dolphin, Konqueror)
 * prostsza instalacja/deinstalacja programu przy kompilacji ze zrodel

2008-03-01 v0.1.3:
 * skanowanie katalogow i dopasowywanie napisow do znalezionych filmow
 * rozwiazanie problemu z pozycja okienek dialogowych pod compizem
 * odswiezona ikona programu
 * balonikowe notyfikacje w tray-u
 * pokazywanie okna do otwierania plikow z filmami przy kliknieciu
   LPM na ikone w tray-u

src/napi.cpp
/*************************************************************************

    QNapi
    Copyright (C) 2008 Krzemin <pkrzemin@o2.pl>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

*************************************************************************/

#include "napi.h"

// Zwraca sume md5 z pierwszych max_size bajtow z pliku
// lub calego, jesli max_size == 0
QString napiFileMd5Sum(const QString & filename, unsigned long max_size)
{
	QFile file(filename);
	QByteArray fileArray;

	file.open(QIODevice::ReadOnly);
	fileArray = ( max_size == 0 ) ? file.readAll() : file.read(max_size);
	file.close();

	QByteArray b = QCryptographicHash::hash(fileArray, QCryptographicHash::Md5);
	fileArray.clear();

	QString out;
	char next[3];

	for(int i = 0; i < 16; i++)
	{
		sprintf(next, "%.2x", (unsigned char)b[i]);
		out += next;
	}

	return out;
}

// Tajemnicza funkcja f() :D
QString napiFDigest(QString input)
{
	if(input.size() != 32) return "";

	int idx[] = {0xe, 0x3, 0x6, 0x8, 0x2},
		mul[] = {2, 2, 5, 4, 3},
		add[] = {0x0, 0xd, 0x10, 0xb, 0x5},
		a, m, i, t, v;

	QString b;
	char tmp[2] = { 0, 0 };

	for(int j = 0; j <= 4; j++)
	{
		a = add[j];
		m = mul[j];
		i = idx[j];

		tmp[0] = input[i].toAscii();
		t = a + (int)(strtol(tmp, NULL, 16));
		v = (int)(strtol(input.mid(t, 2).toLocal8Bit(), NULL, 16));

		sprintf(tmp, "%x", (v * m) % 0x10);
		b += tmp;
	}

	return b;
}

// Zasysa napisy o sumie md5sum w jezyku language i zapisuje je w pliku path
bool napiDownload(const QString & md5sum, const QString & path, const QString & language,
							const QString & nick, const QString & pass)
{
	SyncHTTP http;
	QString urlTxt = napiDownloadUrlTpl.arg(language).arg(md5sum).arg(napiFDigest(md5sum)).arg(nick).arg(pass);

	QUrl url(urlTxt);

	http.setHost(url.host());
	http.syncGet(url.path() + "?" + url.encodedQuery());

	QByteArray buffer = http.readAll();

	if(buffer.indexOf("NPc") == 0)
		return false;

	QFile file(path);
	if(file.exists()) file.remove();
	if(!file.open(QIODevice::WriteOnly))
		return false;

	int r = file.write(buffer);
	file.close();
	return r;
}

// Dopasowuje napisy i usuwa pliki tymczasowe
bool napiMatchSubtitles(const QString & md5sum, const QString & zip_file, const QString & movie_file,
						bool noBackup, const QString & tmp_path, const QString & p7zip_path)
{
	if(!QFile::exists(zip_file)) return false;
	if(!QFile::exists(movie_file)) return false;
	QFileInfo fi(p7zip_path);
	if(!fi.isExecutable()) return false;

	QStringList args;
	args << "x" << "-y" << ("-p"+napiZipPassword) << ("-o"+tmp_path) << zip_file;

	QProcess p7zip;
	p7zip.start(p7zip_path, args);

	if(!p7zip.waitForFinished()) return false;

	QString subtitleFile = tmp_path + "/" + md5sum + ".txt";
	if(!QFile::exists(subtitleFile)) return false;

	QFileInfo mf(movie_file);

	QString newName = mf.path() + "/" + mf.completeBaseName() + ".txt";

	if(QFile::exists(newName))
	{
		if(!noBackup)
		{
			QFileInfo nn(newName);
			QFile::copy(newName, nn.path() + "/" + nn.completeBaseName() + "_old.txt");
		}

		QFile::remove(newName);
	}

	bool r;

	#ifdef Q_WS_WIN
	// Pod windowsem, aby "wyczyscic" atrybuty pliku, tworzymy go na nowo
		QFile f(newName), f2(subtitleFile);
		if(!f.open(QIODevice::WriteOnly) || !f2.open(QIODevice::ReadOnly))
		{
			r = false;
			f.close();
		}
		else
		{
			r = f.write(f2.readAll()) > 0;
			f2.close();
			f.close();
		}
	#else
	// pod normalnymi OS-ami nie trzeba sie gimnastykowac z atrybutami
		r = QFile::copy(subtitleFile, newName);
	#endif

	QFile::remove(subtitleFile);
	QFile::remove(zip_file);

	return r;
}

// Sprawdza uzytkownika w bazie
bool napiCheckUser(const QString & nick, const QString & pass)
{
	SyncHTTP http;
	QString urlTxt = napiCheckUserUrlTpl.arg(nick).arg(pass);

	QUrl url(urlTxt);
	http.setHost(url.host());
	http.syncGet(url.path() + "?" + url.encodedQuery());

	QString buffer = http.readAll();
	if(buffer.indexOf("ok") == 0) return true;
	return false;
}

// Wysyla napisy do bazy
napiUploadResult napiUploadSubtitles(const QString & movie_file, const QString & subtitles_file,
										const QString & language, const QString & nick,
										const QString & pass, bool correct, const QString & comment,
										const QString & tmp_path, const QString & p7zip_path)
{
	if(!QFile::exists(movie_file) || !QFile::exists(subtitles_file))
		return NAPI_FAIL;

	MovieInfo movieInfo(movie_file);
	if(movieInfo.isErr)
		return NAPI_FAIL;

	unsigned long movie_size = QFileInfo(movie_file).size();
	QString movie_md5 = napiFileMd5Sum(movie_file, NAPI_10MB);
	QString subtitles_md5 = napiFileMd5Sum(subtitles_file, NAPI_10MB);

	QString newSubtitlesName = tmp_path + "/" + movie_md5 + ".txt";
	if(QFile::exists(newSubtitlesName)) QFile::remove(newSubtitlesName);
	if(!QFile::copy(subtitles_file, newSubtitlesName))
		return NAPI_FAIL;

	QString zipFileName = QFileInfo(newSubtitlesName).path() + "/"
							+ QFileInfo(newSubtitlesName).completeBaseName() + ".zip";

	QStringList args;
	args << "a" << "-l" << "-y" << zipFileName << ("-p"+napiZipPassword) << newSubtitlesName;

	QProcess p7zip;
	p7zip.start(p7zip_path, args);

	if(!p7zip.waitForFinished())
		return NAPI_FAIL;

	QFile::remove(newSubtitlesName);

	if(!QFile::exists(zipFileName))
		return NAPI_FAIL;

	// Przygotowujemy dane do zapytania POST
	MultipartData postData;

	postData.addBoundary();
	postData.addContentDisposition("name=\"kmt\"");
	postData.addData(comment);

	postData.addBoundary();
	postData.addContentDisposition("name=\"poprawka\"");
	postData.addData(QString(correct ? "true" : "false"));

	postData.addBoundary();
	postData.addContentDisposition("name=\"v\"");
	postData.addData(QString("other"));

	postData.addBoundary();
	postData.addContentDisposition("name=\"l\"");
	postData.addData(language);

	postData.addBoundary();
	postData.addContentDisposition("name=\"t\"");
	postData.addData(napiFDigest(movie_md5));

	postData.addBoundary();
	postData.addContentDisposition("name=\"m_filename\"");
	postData.addData(QFileInfo(movie_file).fileName());

	postData.addBoundary();
	postData.addContentDisposition("name=\"nick\"");
	postData.addData(nick);

	postData.addBoundary();
	postData.addContentDisposition("name=\"pass\"");
	postData.addData(pass);

	postData.addBoundary();
	postData.addContentDisposition("name=\"s_hash\"");
	postData.addData(subtitles_md5);

	postData.addBoundary();
	postData.addContentDisposition("name=\"MAX_FILE_SIZE\"");
	postData.addData(QString("512000"));

	postData.addBoundary();
	postData.addContentDisposition("name=\"plik\"; filename=\"" +
		QFileInfo(zipFileName).fileName() + "\"");
	postData.addContentType("subtitles/zip");

	QFile fZip(zipFileName);
	if(!fZip.open(QIODevice::ReadOnly))
		return NAPI_FAIL;

	postData.addData(fZip.readAll());
	fZip.close();
	QFile::remove(zipFileName);

	postData.addEndingBoundary();

	QByteArray data = postData.requestStream();

	QString movie_fps = QString::number((int)ceil(movieInfo.fps * 100));
	movie_fps.insert(2, ',');

	QString urlTxt = napiUploadUrlTpl.arg(movieInfo.time).arg(movieInfo.width).arg(movieInfo.height)
						.arg(movie_fps).arg(movie_md5).arg(movie_size);

	QUrl url(urlTxt);

	QHttpRequestHeader header("POST", url.path() + "?" + url.encodedQuery());

	header.setValue("Host", url.host());
	header.setValue("Accept", "text/html, */*");
	header.setValue("Content-Type", "multipart/form-data; boundary=" + postData.boundaryTxt());
	header.setValue("Connection", "keep-alive");
	header.setValue("User-Agent", QString("QNapi ") + QNAPI_VERSION);

	SyncHTTP http;
	http.setHost(url.host());
	if(!http.syncRequest(header, data))
		return NAPI_FAIL;

	QString response = http.readAll();

	if(response.indexOf("NPc0") == 0)
		return NAPI_ADDED_NEW;
	if((response.indexOf("NPc2") == 0) || (response.indexOf("NPc3") == 0))
		return NAPI_OK;
	return NAPI_UNKNOWN;
}

// Zglasza raport o niepasujacych napisach
napiReportResult napiReportBad(const QString & movie_file, const QString & language,
								const QString & nick, const QString & pass, const QString & comment,
								QString *response)
{
	QString subtitles_file = QFileInfo(movie_file).path() + "/"
								+ QFileInfo(movie_file).completeBaseName() + ".txt";

	if(!QFile::exists(subtitles_file))
		return NAPI_NO_SUBTITLES;

	MultipartData postData;

	postData.addBoundary();
	postData.addContentDisposition("name=\"nick\"");
	postData.addData(nick);

	postData.addBoundary();
	postData.addContentDisposition("name=\"pass\"");
	postData.addData(pass);

	postData.addBoundary();
	postData.addContentDisposition("name=\"l\"");
	postData.addData(language);

	postData.addBoundary();
	postData.addContentDisposition("name=\"md5\"");
	postData.addData(napiFileMd5Sum(movie_file, NAPI_10MB));

	postData.addBoundary();
	postData.addContentDisposition("name=\"kmt\"");
	postData.addData(comment);
	postData.addEndingBoundary();

	QByteArray data = postData.requestStream();

	QUrl url(napiReportBadUrlTpl);

	QHttpRequestHeader header("POST", url.path());

	header.setValue("Host", url.host());
	header.setValue("Accept", "text/html, */*");
	header.setValue("Content-Type", "multipart/form-data; boundary=" + postData.boundaryTxt());
	header.setValue("Connection", "keep-alive");
	header.setValue("User-Agent", QString("QNapi ") + QNAPI_VERSION);

	SyncHTTP http;
	http.setHost(url.host());
	if(!http.syncRequest(header, data))
		return NAPI_NOT_REPORTED;

	*response = QTextCodec::codecForName("windows-1250")->toUnicode(http.readAll());

	return NAPI_REPORTED;
}

// Konwertuje napisy z jednego kodowania na inne
bool napiConvertFile(const QString & file, const QString & enc_from, const QString & enc_to)
{
	QFile f(file);
	if(!f.open(QIODevice::ReadOnly))
		return false;

	QByteArray fileContent = f.readAll();
	QString contentStr = QTextCodec::codecForName(qPrintable(enc_from))->toUnicode(fileContent);
	f.close();

	if(enc_to.compare("UTF-8", Qt::CaseInsensitive))
	{
		fileContent = QTextCodec::codecForName(qPrintable(enc_to))
						->fromUnicode(contentStr.constData(), contentStr.size());
	}

	if(!f.open(QIODevice::WriteOnly))
		return false;

	f.write(fileContent);
	f.close();

	return true;
}

// Robi to samo co powyzsza funkcja, z tym ze stara sie automatycznie wykryc kodowanie
// zrodlowe konwertowanego pliku
bool napiConvertFile(const QString & file, const QString & enc_to)
{
	QString enc_from;
	QStringList codecs;
	// Tylko takie kodowania obsługuja polskie litery
	codecs << "windows-1250" << "windows-1257" << "ISO-8859-2"
			<< "ISO-8859-13" << "ISO-8859-16" << "UTF-8";

	QFile f(file);
	if(!f.open(QIODevice::ReadOnly))
		return false;

	QByteArray testData = f.read(5000);

	f.close();

	foreach(QString codec, codecs)
	{
		QTextStream ts(testData);
		ts.setCodec(qPrintable(codec));
		QString encodedData = ts.readAll();
		QStringList chars = QString::fromUtf8("ą/ś/ż/ć/ń/ł/ó/ę").split("/");

		int i;
		for (i = 0; i < chars.count(); i++)
		{
			if(!encodedData.contains(chars[i], Qt::CaseInsensitive))
				break;
		}

		if(i == chars.count())
		{
			enc_from = codec;
			break;
		}
	}

	if(enc_from.isEmpty())
		return false;

	return napiConvertFile(file, enc_from, enc_to);
}

// Zakladanie kont uzytkownikow na serwerze napi
bool napiCreateUser(const QString & nick, const QString & pass, const QString & email,
					QString * response)
{
	MultipartData postData;

	postData.addBoundary();
	postData.addContentDisposition("name=\"login\"");
	postData.addData(nick);

	postData.addBoundary();
	postData.addContentDisposition("name=\"haslo\"");
	postData.addData(pass);

	postData.addBoundary();
	postData.addContentDisposition("name=\"mail\"");
	postData.addData(email);

	postData.addBoundary();
	postData.addContentDisposition("name=\"z_programu\"");
	postData.addData(QString("true"));

	postData.addEndingBoundary();

	QByteArray data = postData.requestStream();

	QUrl url(napiCreateUserUrlTpl);

	QHttpRequestHeader header("POST", url.path());

	header.setValue("Host", url.host());
	header.setValue("Accept", "text/html, */*");
	header.setValue("Content-Type", "multipart/form-data; boundary=" + postData.boundaryTxt());
	header.setValue("Connection", "keep-alive");
	header.setValue("User-Agent", QString("QNapi ") + QNAPI_VERSION);

	SyncHTTP http;
	http.setHost(url.host());
	if(!http.syncRequest(header, data))
		return false;

	*response = QTextCodec::codecForName("windows-1250")->toUnicode(http.readAll());

	return true;
}

// Sprawdza poprawnosc sciezki do 7zipa (z konfiguracji)
bool napiCheck7Zip()
{
	return QFileInfo(GlobalConfig().p7zipPath()).isExecutable();
}

// Sprawdza poprawnosc sciezki do katalogu tymczasowego (z konfiguracji)
bool napiCheckTmpPath()
{
	QFileInfo f(GlobalConfig().tmpPath());
	return f.isDir() && f.isWritable();
}
